### PR TITLE
Refs #7072 Added XTYPES_EXCEPTION flag to activate runtime exceptions

### PR DESCRIPTION
**Pull request by LuisGP**
This PR allows the user to enable runtime exceptions instead of xtypes_asserts.
These exceptions will be launch at release too and allow protection when the programmer has no control of the data until runtime.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/27/)

File: test/unitary/parser/parser_test.cpp [hunk 1]
```diff
@@ -20,6 +20,37 @@
 using namespace eprosima::xtypes;
 using namespace eprosima::xtypes::idl;
 
+#if defined(XTYPES_EXCEPTIONS)
+#define ASSERT_OR_EXCEPTION(exp, msg)                                                                       \
+{                                                                                                           \
+            try                                                                                             \
+            {                                                                                               \
+                { exp }                                                                                     \
```
Comment (lemunozm) [inline test/unitary/parser/parser_test.cpp]: Why the ´{  }´, I am curious

File: test/unitary/parser/parser_test.cpp [hunk 2]
```diff
@@ -183,7 +214,7 @@ TEST (IDLParser, array_sequence_struct_test)
         }
         else
         {
-            ASSERT_DEATH(data["my_char6_seq"][i].push(static_cast<char>(i)), "Assertion");
+            ASSERT_OR_EXCEPTION(data["my_char6_seq"][i].push(static_cast<char>(i));, "out of bounds");
```
Comment (lemunozm) [inline test/unitary/parser/parser_test.cpp:187]: I think it is dangerous if the user needs to put the `;` here.

File: test/unitary/xtypes/struct_type.cpp [hunk 3]
```diff
@@ -19,6 +19,38 @@
 #include <cmath>
 #include <bitset>
 
+#if defined(XTYPES_EXCEPTIONS)
```
Comment (lemunozm) [inline test/unitary/xtypes/struct_type.cpp]: Why is it copy in several files instead of including a header only for this purpose?

**Review by lemunozm — CHANGES_REQUESTED** [review]

File: test/unitary/parser/parser_test.cpp (hunk 1 above)
Comment (LuisGP) [reply test/unitary/parser/parser_test.cpp]: It's just an isolation guard.

File: test/unitary/parser/parser_test.cpp (hunk 2 above)
Comment (LuisGP) [reply test/unitary/parser/parser_test.cpp:187]: It is not supposed to be used by the user, only for our testing.

File: test/unitary/xtypes/struct_type.cpp (hunk 3 above)
Comment (LuisGP) [reply test/unitary/xtypes/struct_type.cpp]: Yes, but I thought it isn't worth enough in this case.

File: test/unitary/parser/parser_test.cpp (hunk 2 above)
Comment (lemunozm) [reply test/unitary/parser/parser_test.cpp:187]: But what if a user read this? can be confused?

Comment (LuisGP) [reply test/unitary/parser/parser_test.cpp:187]: Do you think a user is going to implement a test?
Maybe adding an example of usage of the assertion or the exception would be useful.

Comment (lemunozm) [reply test/unitary/parser/parser_test.cpp:187]: Added an example of exception usage.

File: test/unitary/xtypes/struct_type.cpp (hunk 3 above)
Comment (lemunozm) [reply test/unitary/xtypes/struct_type.cpp]: It is worth it :D

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/28/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/29/)

File: test/unitary/parser/parser_test.cpp [hunk 4]
```diff
@@ -20,6 +20,37 @@
 using namespace eprosima::xtypes;
 using namespace eprosima::xtypes::idl;
 
+#if defined(XTYPES_EXCEPTIONS)
```
Comment (lemunozm) [inline test/unitary/parser/parser_test.cpp]: Remove it and include utils :D

File: examples/exceptions_asserts.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/**
+* This example shows how to make use of the builtin assert and exception system.
+*/
+
+#include <xtypes/xtypes.hpp>
+#include <iostream>
+
+// Auxiliar define to manage both assert (debug only) and exceptions (both)
+#if defined(XTYPES_EXCEPTIONS)
```
Comment (lemunozm) [inline examples/exceptions_asserts.cpp:9]: Remove it and include utils :D

**Review by lemunozm — CHANGES_REQUESTED** [review]

File: examples/exceptions_asserts.cpp (hunk 5 above)
Comment (LuisGP) [reply examples/exceptions_asserts.cpp:9]: It's not the same define here. It is full of comments and *couts* to show information ;).

**Review by lemunozm — APPROVED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/31/)

**Pull request merged**